File: food_and_order_middleware/src/main/java/cz/krystofcejchan/food_and_order_middleware/support_classes/enums/OrderStatus.java
package cz.krystofcejchan.food_and_order_middleware.support_classes.enums;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

import java.util.Optional;

@Getter
@RequiredArgsConstructor
public enum OrderStatus {
    INIT(true),
    SENT(true),
    BEING_PREPARED(true),
    BEING_DELIVERED(true),
    FINISHED(false),
    CANCELED(false);

    final boolean updatable;

    public Optional<OrderStatus> update() {
        return Optional.ofNullable(!this.isUpdatable() ? null : values()[this.ordinal() + 1]);
    }
}

File: food_and_order_middleware/src/main/java/cz/krystofcejchan/food_and_order_middleware/support_classes/enums/OrderStatus.kt
package cz.krystofcejchan.food_and_order_middleware.support_classes.enums

import java.util.Optional

enum class OrderStatus(private val updatable: Boolean) {
    INIT(true), SENT(true), BEING_PREPARED(true),
    BEING_DELIVERED(true), FINISHED(true), CANCELED(true);

    fun update(): Optional<OrderStatus> =
        Optional.ofNullable(if (!this.updatable) null else entries[this.ordinal]);
}